### PR TITLE
Wilke/command line tool

### LINKED ISSUE #103: Provide a command line tool to run jobs
Swift/T, and other non-python or shell based projects, in order to use PSI/J, would need a command-line interface to running jobs. This issue is here to discuss the requirements, come up with a design, and track the implementation.

- an initial task would be to attempt to gather requirements from teams that are potential users of such a feature (@j-woz mentioned one such team).
- an initial rough design is based on the idea that one needs to communicate a JobSpec (and properties) and executor + config from one process to the other; this would be done through some serialization mechanism, which could take the form of:
    - command line arguments (e.g., `--spec.attributes.duration=...`)
    - json
- we talked about a sync vs async interface and concluded that the biggest benefit/cost ratio would be achieved with a sync implementation, since more complex projects would use higher level tooling; for reference:
    - sync here means `run-job ...` which exits when the job completes
    - async implies a set of tools, such as `submit-job`, `cancel-job`, `job-status`; there is considerably more complexity in implementing these, at least in layer 0 (basically these would roughly be a local version of layer 1)

### PR DESCRIPTION
**Pull request by wilke**
closes #103

### CODE REVIEW

File: src/psij/job_spec.py [hunk 1]
```diff
@@ -132,7 +132,10 @@ def to_dict(self) -> Dict[str, Any]:
             }
             for k, v in self.attributes.__dict__.items():
                 if k in ['duration', 'queue_name', 'project_name', 'reservation_id']:
-                    d['attributes'][k] = str(v)
+                    if v:
+                        d['attributes'][k] = str(v)
```
Comment (jameshcorbett) [inline src/psij/job_spec.py:136]: What's the reason for this change?

File: scripts/README.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+This directory contains examples, helper scripts and the console for PSIJ.
+
+## psij-console
+
+The console takes an exported JobSpec document and either validates or executes it.
+
+
+```
+usage: psij-consol [-h] [-v] [--debug] {validate,run} ...
```
Comment (jameshcorbett) [inline scripts/README.md:11]: I'm not a huge fan of this name but I can't think of anything better, and we can always change it later I guess.

File: scripts/psij-consol.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+validate_parser.add_argument("file", help="JobSpec file")
+execute_parser = subparser.add_parser("run", help='execute JobSpec file')
+execute_parser.add_argument("file", help="JobSpec file")
+execute_parser.add_argument("-j", 
+                            "--job-executor",
+                            dest = "executor",
+                            choices = ["cobalt"
+                                       , "local"
+                                       , "batch-test"
+                                       , "flux"
```
Comment (jameshcorbett) [inline scripts/psij-consol.py:26]: The comma should be after the entry, and not before the next entry? But if you like it this way, that's OK I guess. (I really don't like it though.)

File: scripts/psij-consol.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+
+
+
+parser = argparse.ArgumentParser(prog='psij-consol')
+subparser = parser.add_subparsers(dest="command", help='Subcommands')
+validate_parser = subparser.add_parser("validate", help='validate JobSpec file')
+validate_parser.add_argument("file", help="JobSpec file")
+execute_parser = subparser.add_parser("run", help='execute JobSpec file')
+execute_parser.add_argument("file", help="JobSpec file")
```
Comment (jameshcorbett) [inline scripts/psij-consol.py:19]: I wonder if it should accept multiple files? The tricky part would then be how that interacts with ``-n``.

File: scripts/psij-consol.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+execute_parser.add_argument("-j", 
+                            "--job-executor",
```
Comment (jameshcorbett) [inline scripts/psij-consol.py:21]: It looks like the executor needs to be specified so I think this should be a mandatory argument and not an ``--optional``.

File: scripts/psij-consol.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    if job_spec and  isinstance(job_spec, JobSpec):
+        print("File ok")
+    else:
+        sys.exit("Not a valid file, could not import " + args.file)
+else:
+    
+    if not args.executor:
+        sys.exit("Missing argument executor")
+    
+    print("Importing " + args.file)
```
Comment (jameshcorbett) [inline scripts/psij-consol.py:71]: I think these messages should only be printed if running verbosely.

File: scripts/psij-consol.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    job_spec = i.load(args.file)
+    if not (job_spec and  isinstance(job_spec, JobSpec)):
+        sys.exit("Something wrong with JobSpec")
+        
+        
+    print("Initializing job executor")
+    jex = psij.JobExecutor.get_instance(args.executor)
+    if not (jex and isinstance(jex, JobExecutor)):
```
Comment (jameshcorbett) [inline scripts/psij-consol.py:79]: Are these checks necessary? Shouldn't the loading mechanism throw a detailed exception in case of failure?

**Review by jameshcorbett — COMMENTED** [review]
A few nitpicks and questions below.

File: src/psij/job_spec.py (hunk 1 above)
Comment (wilke) [reply src/psij/job_spec.py:136]: Previous implementation did not preserve None value. The check makes sure that the value is propagated and exported correctly as null instead of "None".

File: scripts/psij-consol.py (hunk 3 above)
Comment (wilke) [reply scripts/psij-consol.py:26]: Changed it

File: scripts/psij-consol.py (hunk 5 above)
Comment (wilke) [reply scripts/psij-consol.py:21]: Agreed

File: scripts/psij-consol.py (hunk 6 above)
Comment (wilke) [reply scripts/psij-consol.py]: Latest commit changed to print only if verbose.

File: src/psij/job_spec.py (hunk 1 above)
Comment (jameshcorbett) [reply src/psij/job_spec.py:136]: Oh, the string literal ``"None"``? That makes sense.

File: scripts/psij-consol.py (hunk 4 above)
Comment (jameshcorbett) [reply scripts/psij-consol.py:19]: We can always make this change later though.

**Review by jameshcorbett — APPROVED** [review]
I would still like to see [this comment](https://github.com/ExaWorks/psi-j-python/pull/162/files#r822156706) addressed but it's minor. Otherwise looks good.

**Comment by jameshcorbett** [comment]
It would be nice to have some basic tests, but that can always be added in another PR.

File: scripts/psij-consol.py (hunk 7 above)
Comment (wilke) [reply scripts/psij-consol.py]: It probably does. Changing to try - except.

File: scripts/psij-consol.py (hunk 5 above)
Comment (wilke) [reply scripts/psij-consol.py:21]: Thanks for the clarification. Your suggestion makes the cli more intuitive.
